src/Renci.SshNet/Common/PipeInputStream.cs

```diff
@@ -0,0 +1,112 @@
+﻿namespace Renci.SshNet.Common 
+{
+    using System;
+    using System.IO;
+
+    internal class PipeInputStream : Stream
+    {
+        private LinkedListQueue<byte[]> _queue;
+        private byte[] _current;
+        private int _currentPosition;
+        private bool _isDisposed;
+
+        public PipeInputStream(LinkedListQueue<byte[]> queue)
+        {
+            _queue = queue;
+        }
+
+        public override void Flush()
+        {
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+                throw new ArgumentNullException("buffer");
+            if (offset + count > buffer.Length)
+                throw new ArgumentException("The sum of offset and count is greater than the buffer length.");
+            if (offset < 0 || count < 0)
+                throw new ArgumentOutOfRangeException("offset", "offset or count is negative.");
+            if (_isDisposed)
+                throw CreateObjectDisposedException();
+
+            if (_current == null || _currentPosition == _current.Length)
+            {
+                if (_queue.IsCompleted || !_queue.TryTake(out _current))
+                    return 0;
+
+                _currentPosition = 0;
+            }
+
+            var avail = _current.Length - _currentPosition;
+            if (count > avail)
+                count = avail;
+
+            Buffer.BlockCopy(_current, _currentPosition, buffer, offset, count);
+
+            _currentPosition += count;
+            return count;
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override bool CanRead
+        {
+            get { return true; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return false; }
```

Comment:
Correct, I didn't notice it's in the Stream spec to return false if closed/disposed.